File: src/dashboard/src/types/workflow.ts
export type WorkflowStatus = 'active' | 'inactive' | 'draft'
export type WorkflowRunStatus = 'queued' | 'running' | 'completed' | 'failed' | 'cancelled'
export type JobStatus = 'queued' | 'running' | 'success' | 'failure' | 'skipped' | 'cancelled'
export type StepStatus = 'pending' | 'running' | 'success' | 'failure' | 'skipped'
export type TriggerType = 'manual' | 'push' | 'pull_request' | 'schedule' | 'webhook' | 'merge'
export type RunnerType = 'local' | 'docker'

export interface WorkflowStepDef {
  name: string
  run?: string
  uses?: string
  with?: Record<string, unknown>
  env?: Record<string, string>
  if?: string
  continue_on_error?: boolean
  timeout_minutes?: number
}

export interface WorkflowJobDef {
  name?: string
  runs_on?: RunnerType
  needs?: string[]
  steps: WorkflowStepDef[]
  matrix?: Record<string, string[]>
  environment?: string
  if?: string
  env?: Record<string, string>
  timeout_minutes?: number
}

export interface WorkflowDefinition {
  name: string
  description?: string
  on?: Record<string, unknown>
  env?: Record<string, string>
  jobs: Record<string, WorkflowJobDef>
}

export interface Workflow {
  id: string
  name: string
  description: string
  status: WorkflowStatus
  project_id: string | null
  definition: WorkflowDefinition
  yaml_content: string | null
  version: number
  created_by: string | null
  created_at: string
  updated_at: string
  last_run_at: string | null
  last_run_status: WorkflowRunStatus | null
}

export interface WorkflowStep {
  id: string
  job_id: string
  name: string
  status: StepStatus
  run?: string
  uses?: string
  output?: string
  error?: string
  duration_ms?: number
  started_at?: string
  completed_at?: string
}

export interface WorkflowJob {
  id: string
  run_id: string
  name: string
  status: JobStatus
  runner: RunnerType
  needs: string[]
  environment?: string
  started_at?: string
  completed_at?: string
  duration_seconds?: number
  steps: WorkflowStep[]
}

export interface WorkflowRun {
  id: string
  workflow_id: string
  workflow_name: string
  trigger_type: TriggerType
  trigger_payload: Record<string, unknown>
  status: WorkflowRunStatus
  started_at: string
  completed_at?: string
  duration_seconds?: number
  total_cost: number
  error_summary?: string
  jobs: WorkflowJob[]
}

export interface WorkflowLog {
  timestamp: string
  level: string
  message: string
}

// Phase 2: Trigger & Schedule types
export interface CronSchedule {
  cron: string
  timezone: string
  is_active: boolean
  next_run?: string
}

export interface WebhookConfig {
  id: string
  workflow_id: string
  url: string
  secret?: string
  is_active: boolean
  allowed_events: string[]
  created_at: string
}

export interface SecretVariable {
  id: string
  name: string
  scope: 'workflow' | 'project' | 'global'
  scope_id?: string
  created_by?: string
  created_at: string
  updated_at: string
}

// Phase 3: Artifact & Template types
export interface RetryConfig {
  max_attempts: number
  backoff: 'linear' | 'exponential'
  delay_seconds?: number
}

export interface Artifact {
  id: string
  run_id: string
  job_id?: string
  step_id?: string
  name: string
  path: string
  size_bytes: number
  content_type: string
  retention_days: number
  expires_at?: string
  created_at: string
}

export type TemplateCategory = 'ci' | 'deploy' | 'test' | 'utility'

export interface Template {
  id: string
  name: string
  description: string
  category: TemplateCategory
  tags: string[]
  yaml_content: string
  icon: string
  popularity: number
  created_at: string
  updated_at: string
}

File: src/backend/models/workflow.py
"""Workflow automation models."""

from datetime import datetime
from enum import Enum
from typing import Any

from pydantic import BaseModel, Field


class WorkflowStatus(str, Enum):
    ACTIVE = "active"
    INACTIVE = "inactive"
    DRAFT = "draft"


class WorkflowRunStatus(str, Enum):
    QUEUED = "queued"
    RUNNING = "running"
    COMPLETED = "completed"
    FAILED = "failed"
    CANCELLED = "cancelled"


class JobStatus(str, Enum):
    QUEUED = "queued"
    RUNNING = "running"
    SUCCESS = "success"
    FAILURE = "failure"
    SKIPPED = "skipped"
    CANCELLED = "cancelled"


class StepStatus(str, Enum):
    PENDING = "pending"
    RUNNING = "running"
    SUCCESS = "success"
    FAILURE = "failure"
    SKIPPED = "skipped"


class TriggerType(str, Enum):
    MANUAL = "manual"
    PUSH = "push"
    PULL_REQUEST = "pull_request"
    SCHEDULE = "schedule"
    WEBHOOK = "webhook"
    MERGE = "merge"


class RunnerType(str, Enum):
    LOCAL = "local"
    DOCKER = "docker"


# Pydantic schemas for API
class RetryConfig(BaseModel):
    max_attempts: int = 1
    backoff: str = "linear"  # linear, exponential
    delay_seconds: float = 1.0


class WorkflowStepDef(BaseModel):
    name: str
    id: str | None = None  # step id for output referencing
    run: str | None = None
    uses: str | None = None
    with_args: dict[str, Any] | None = Field(None, alias="with")
    env: dict[str, str] | None = None
    if_condition: str | None = Field(None, alias="if")
    continue_on_error: bool = False
    timeout_minutes: int = 30
    retry: RetryConfig | dict | None = None
    outputs: dict[str, str] | None = None  # output definitions


class WorkflowJobDef(BaseModel):
    name: str | None = None
    runs_on: RunnerType = RunnerType.LOCAL
    needs: list[str] = Field(default_factory=list)
    steps: list[WorkflowStepDef] = Field(default_factory=list)
    matrix: dict[str, list[str]] | None = None
    matrix_exclude: list[dict[str, str]] | None = None
    matrix_include: list[dict[str, str]] | None = None
    environment: str | None = None
    if_condition: str | None = Field(None, alias="if")
    env: dict[str, str] | None = None
    timeout_minutes: int = 60
    outputs: dict[str, str] | None = None


class TriggerConfig(BaseModel):
    push: dict[str, Any] | None = None
    pull_request: dict[str, Any] | None = None
    schedule: list[dict[str, str]] | None = None
    manual: dict[str, Any] | None = None
    webhook: dict[str, Any] | None = None


class WorkflowDefinitionSchema(BaseModel):
    name: str
    description: str = ""
    on: TriggerConfig = Field(default_factory=TriggerConfig)
    env: dict[str, str] = Field(default_factory=dict)
    jobs: dict[str, WorkflowJobDef] = Field(default_factory=dict)


# API Request/Response schemas
class WorkflowCreate(BaseModel):
    name: str
    description: str = ""
    yaml_content: str | None = None
    definition: WorkflowDefinitionSchema | None = None
    project_id: str | None = None


class WorkflowUpdate(BaseModel):
    name: str | None = None
    description: str | None = None
    yaml_content: str | None = None
    definition: WorkflowDefinitionSchema | None = None
    status: WorkflowStatus | None = None
    project_id: str | None = None


class WorkflowResponse(BaseModel):
    id: str
    name: str
    description: str
    status: WorkflowStatus
    project_id: str | None
    definition: dict
    yaml_content: str | None
    version: int
    created_by: str | None
    created_at: datetime
    updated_at: datetime
    last_run_at: datetime | None = None
    last_run_status: WorkflowRunStatus | None = None


class WorkflowRunTrigger(BaseModel):
    trigger_type: TriggerType = TriggerType.MANUAL
    inputs: dict[str, Any] = Field(default_factory=dict)
    branch: str | None = None


class WorkflowRunResponse(BaseModel):
    id: str
    workflow_id: str
    workflow_name: str
    trigger_type: TriggerType
    trigger_payload: dict
    status: WorkflowRunStatus
    started_at: datetime
    completed_at: datetime | None
    duration_seconds: float | None
    total_cost: float
    error_summary: str | None
    jobs: list["WorkflowJobResponse"] = Field(default_factory=list)


class WorkflowJobResponse(BaseModel):
    id: str
    run_id: str
    name: str
    status: JobStatus
    runner: RunnerType
    needs: list[str]
    environment: str | None
    started_at: datetime | None
    completed_at: datetime | None
    duration_seconds: float | None
    steps: list["WorkflowStepResponse"] = Field(default_factory=list)


class WorkflowStepResponse(BaseModel):
    id: str
    job_id: str
    name: str
    status: StepStatus
    run: str | None
    uses: str | None
    output: str | None
    error: str | None
    duration_ms: int | None
    started_at: datetime | None
    completed_at: datetime | None


class WorkflowListResponse(BaseModel):
    workflows: list[WorkflowResponse]
    total: int


class WorkflowRunListResponse(BaseModel):
    runs: list[WorkflowRunResponse]
    total: int
